graphula-core/src/Graphula.hs

```diff
@@ -429,85 +446,156 @@ class HasDependencies a where
         (toEot a)
         (toEot dependencies)
 
-data GenerationFailure =
-  GenerationFailureMaxAttempts TypeRep
+data GenerationFailure
+  = GenerationFailureMaxAttemptsToConstrain TypeRep
+  | GenerationFailureMaxAttemptsToInsert TypeRep
   deriving (Show, Typeable, Eq)
 
 instance Exception GenerationFailure
 
-type GraphulaNode m a =
-  ( Generate m a
-  , HasDependencies a
-  , Logging m a
-  , PersistEntityBackend a ~ SqlBackend
-  , PersistEntity a
-  , Typeable a
-  , EntityKeyGen a
-  )
+type GraphulaNode m a
+  = ( Generate m a
+    , HasDependencies a
+    , Logging m a
+    , PersistEntityBackend a ~ SqlBackend
+    , PersistEntity a
+    , Typeable a
+    , EntityKeyGen a
+    )
+
+-- | Options for generating an individual node
+--
+-- @'NodeOptions'@ can be created and combined with the Monoidal
+-- operations @'(<>)'@ and @'mempty'@
+--
+-- @
+-- a1 <- root @A mempty
+-- a2 <- root @A $ keyed $ AKey 1
+-- a3 <- root @A $ edit $ \a -> a { someField = True }
+-- a4 <- root @A $ suchThat $ (== True) . someField
+-- a5 <- root @A $ keyed (AKey 2) <> edit (\a -> a { someField = True })
+-- a6 <- root @A $ keyed (AKey 3) <> do edit $ \a -> a { someField = True }
+-- @
+--
+data NodeOptions a = NodeOptions
+  { nodeOptionsEdit :: Kendo Maybe a
+  , nodeOptionsKeyed :: Last (Gen (Maybe (Key a)))
+  }
+  deriving (Generic)
 
-{-|
-  Generate and edit a value with data dependencies. This leverages
-  'HasDependencies' to insert the specified data in the generated value. All
-  dependency data is inserted after editing operations.
+instance Semigroup (NodeOptions a) where
+  (<>) = gmappend
+  {-# INLINE (<>) #-}
 
-  > nodeEdit @Dog (ownerId, veterinarianId) $ \dog ->
-  >   dog {name = "fido"}
--}
-nodeEditWith
-  :: forall a m
-   . GraphulaContext m '[a]
-  => Dependencies a
-  -> (a -> a)
-  -> m (Entity a)
-nodeEditWith dependencies edits = 10 `attemptsToInsertWith` do
-  x <- (`dependsOn` dependencies) . edits <$> generateNode
-  logNode x
-  pure x
+instance Monoid (NodeOptions a) where
+  mempty = gmempty
+  {-# INLINE mempty #-}
 
-{-|
-  Generate a value with data dependencies. This leverages 'HasDependencies' to
-  insert the specified data in the generated value.
+-- | Like @'Endo'@ but use Kliesli composition
+newtype Kendo m a = Kendo { appKendo :: a -> m a }
 
-  > nodeEdit @Dog (ownerId, veterinarianId)
--}
-nodeWith
-  :: forall a m . GraphulaContext m '[a] => Dependencies a -> m (Entity a)
-nodeWith = flip nodeEditWith id
+instance Monad m => Semigroup (Kendo m a) where
+  (<>) = coerce @((a -> m a) -> (a -> m a) -> a -> m a) (<=<)
+  {-# INLINE (<>) #-}
 
-{-|
-  Generate and edit a value that does not have any dependencies.
+instance Monad m => Monoid (Kendo m a) where
+  mempty = coerce @(a -> m a) pure
```

Comment:
This feels like a "because I can" use of `coerce` where `Kendo pure` is succinct and readable.

Reply:
Every use of `coerce` is a "because I can" :grinning: